Desktop: Keyboard shortcut editor

anjulalk opened the pull request:
Contains the implementation of Keyboard shortcut editor GUI. This PR includes,
* Integration with KeymapService and CommandService
* Display a list of available commands along with their keyboard shortcuts
* Allow editing a keyboard shortcut, or reset it to the default shortcut
* Synchronize changes with the `keymap-desktop.json` file in the profile directory
* Search for keyboard shortcuts by their command name or label

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+	}
+};
+
+const initialKeymap = commandNames.map(commandName => {
+	return {
+		label: CommandService.instance().label(commandName),
+		accelerator: getAccelerator(commandName),
+		command: commandName,
+		isEditing: false,
+	};
+});
```
That should be loaded inside the component, using useEffect for example. In general, only constants or functions can be defined outside the component. Any state-related variables should be defined inside the component (or as hooks).

anjulalk replied:
I've passed a state initialization function to useState(). It'll only be initialized once per lifetime.

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+	};
+});
+
+const KeymapConfigScreen = (props: { theme: Object }) => {
```
Please define an `interface Props` to list the supported properties. `theme` is a number by the way.

anjulalk replied:
> theme is a number by the way.

Thanks for pointing that out. I've corrected it and created separate interfaces for each component.

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+	const [filter, setFilter] = React.useState('');
+
+	const theme = themeStyle(props.theme);
+	const containerStyle = {
+		...theme.containerStyle,
+		padding: 10,
+		overflow: 'auto',
+	};
```
Styles should be defined using `buildStyle()` (see other components for examples)

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+	};
+
+	const updateAccelerator = (commandName: string, accelerator: string) => {
+		const _keymap = [...keymap];
```
For consistency with the rest of the code base, please don't use underscore-variables. By convention, copies of state should be named `newKeymap`

anjulalk replied:
Sorry about that. I'll try to watch out for naming convention mismatches from now on.

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+		setKeymap(_keymap);
+	};
+
+	const updateIsEditing = (commandName: string, isEditing: boolean) => {
+		const _keymap = [...keymap];
+
+		_keymap.forEach(item => item.isEditing = false);
+		_keymap.find(item => item.command === commandName).isEditing = isEditing;
+		setKeymap(_keymap);
+	};
```
You should probably use `useCallback` for this and updateAccelerator otherwise you'll have issues with stale state.

anjulalk replied:
Indeed I was putting it at the risk of having a stale state. As a solution, I passed a function to useState(), which changes state by comparing it to previous state. But I don't understand why useCallback() is neccessary to prevent stale state? I'll be glad if you could point me if there's an issue.

anjulalk replied:
As I understand, the optimization from useCallback() isn't significant in this case (as there's not many elements here) and will turn out to be an overhead?

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+	);
+};
+
+const TopActions = (props: {
```
Please put one component per file. Also please put everything under gui/KeymapConfig/

anjulalk replied:
I've made a folder and put the components in different files. However I included `TopActions` right into `KeymapConfigScreen` because it didn't seem necessary to have it's own component. The shortcut recorder however is defined in it's own file. I've also separated the styles from the component.

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+};
+
+const KeymapTable = (props: {
+	theme: any,
```
Please don't use `any` when the property has a specific type, in this case `number`. Also likewise all props should be defined in an interface.

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+			<td style={theme.textStyle}>
+				{
+					item.isEditing ?
+						<ShortcutRecorder
+							updateAccelerator={(accelerator: string) => props.updateAccelerator(item.command, accelerator)}
+							theme={props.theme}
+						/>
+						:
+						<div onClick={() => props.updateIsEditing(item.command, true)}>
+							{item.accelerator.length ? item.accelerator : _('Disabled')}
+						</div>
```
Please define callback handlers in separate functions.

laurent22 commented on ElectronClient/gui/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,229 @@
+	};
+
+
+	const handleKeydown = (
+		e: React.KeyboardEvent<HTMLDivElement>
+	) => {
+		e.preventDefault();
```
Please name the event `event`, like in the rest of the codebase.

laurent22 commented on ReactNativeClient/lib/KeymapUtils.ts:
```diff
@@ -0,0 +1,43 @@
+const { shim } = require('lib/shim');
+
+export const DOMtoElectronAccelerator = (e: React.KeyboardEvent<HTMLDivElement>) => {
```
`domToElectronAccelerator`. Only classes or components should start with an uppercase.

laurent22 commented on ReactNativeClient/lib/KeymapUtils.ts:
```diff
@@ -0,0 +1,43 @@
+	return parts.join('+');
+};
+
+export const DOMtoElectronKey = (key: string) => {
```
domToElectronKey

laurent22 commented on ReactNativeClient/lib/KeymapUtils.ts:
```diff
@@ -0,0 +1,43 @@
+};
+
+export const DOMtoElectronKey = (key: string) => {
+	if (/^([A-Z0-9]|F1*[1-9]|F10|F2[0-4]|[`~!@#$%^&*()\-_=[\]\\{}|;':",./<>?]|Enter|Tab|Backspace|Delete|Insert|Home|End|PageUp|PageDown|Escape|MediaStop|MediaPlayPause|PrintScreen])$/.test(key)) return key;
+	if (/^([a-z])$/.test(key)) return key.toUpperCase();
+	if (/^Arrow(Up|Down|Left|Right)|Audio(VolumeUp|VolumeDown|VolumeMute)$/.test(key)) return key.slice(5);
```
I think I've seen this code in KeyMapService. It should only be in one place. In fact I'm not entirely sure you should have a utils class when you already have KeyMapService. It's going to cause duplicate functionalities and code.

anjulalk replied:
It's a bit different. But I refactored the function so that it can use the same RegEx you've mentioned. 

> In fact I'm not entirely sure you should have a utils class when you already have KeyMapService.

I see. I'll be adding these two functions to KeymapService as static functions once that PR gets merged.

laurent22 commented on ReactNativeClient/lib/services/CommandService.ts:
```diff
@@ -243,6 +243,12 @@ export default class CommandService extends BaseService {
 
+	label(commandName:string):string {
+		const command = this.commandByName(commandName);
+		if (!command) return null;
```
Could you make it throw an error here instead of returning null? Normally all commands must be declared early and it's not declared, it means there's a bug somewhere which should be fixed.

anjulalk replied:
I've added a throw to the function.

anjulalk commented:
Thanks a lot for the early feedback. It'll certainly help me move in the right direction. I believe I addressed all the issues you've mentioned.

I'll continue to refine the code from examples as I progress. Thanks!

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,127 @@
+
+import styles_ from './styles';
+import { ShortcutRecorder } from './ShortcutRecorder';
+const CommandService = require('lib/services/CommandService').default;
```
Please always use `import` when importing a TypeScript file into another TypeScript file to benefit from type checking.

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,127 @@
+	);
+};
+
+// Placeholder
+const KeymapService_instance_getAccelerator = (commandName: string) => {
+	switch (commandName) {
+	case 'textCopy':
+		return 'Ctrl+C';
+	case 'textCut':
+		return 'Ctrl+X';
+	case 'textPaste':
+		return 'Ctrl+V';
+	default:
+		throw new Error('Invalid command');
+	}
+};
+
+// Placeholder
+const KeymapService_instance_getCommands = () => [
+	'textCopy',
+	'textCut',
+	'textPaste',
+];
```
I assume this will come from KeymapService later on? (i.e. there shouldn't be a static list of commands in this component).

laurent22 commented on ElectronClient/gui/KeymapConfig/ShortcutRecorder.tsx:
```diff
@@ -0,0 +1,119 @@
+		const newAccelerator = KeymapService_domToElectronAccelerator(event);
+
+		switch (newAccelerator) {
+		case 'Enter':
+			props.updateAccelerator(accelerator);
+			props.updateIsEditing(false);
+			break;
+		case 'Escape':
+			return props.updateIsEditing(false);
+		default:
```
A child component should not directly set data on the parent. Instead please use event handlers: `onSubmit`, `onCancel`. All React components should be controlled ones: https://reactjs.org/docs/forms.html#controlled-components

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,127 @@
+				<input
+					value={filter}
+					onChange={event => setFilter(event.target.value)}
+					placeholder={_('Search..')}
```
"Search..."

laurent22 commented on ElectronClient/gui/KeymapConfig/styles/index.ts:
```diff
@@ -0,0 +1,51 @@
+export default function styles(props: KeymapConfigScreenProps) {
+	return buildStyle('KeymapConfigScreen', props.theme, (theme: any) => {
+		return {
+			containerStyle: {
+				...theme.containerStyle,
+				padding: 15,
+			},
+			topActionsStyle: {
+				display: 'flex',
+				flexDirection: 'row',
+			},
+			inlineButtonStyle: {
+				...theme.buttonStyle,
+				padding: 0,
+				marginLeft: 12,
+			},
+			filterInputStyle: {
```
It's a minor thing but please don't name the style elements as "somethingStyle" since we know it's a style. So "containerStyle" => "container", "topActionsStyle" => "topActions", etc.

laurent22 commented on ReactNativeClient/lib/components/shared/config-shared.js:
```diff
@@ -156,6 +156,12 @@ shared.settingsSections = createSelector(
 			isScreen: true,
 		});
 
+		output.push({
+			name: 'keymap',
+			metadatas: [],
+			isScreen: true,
+		});
```
As this is shared with mobile, please check that mobile still works.

anjulalk replied:
I've checked it and it works without any issue..

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+	insertTemplate: _('Insert template'),
+	zoomActualSize: _('Actual Size'),
+	// Conditionally set the label because it's different in macOS and other platforms
+	config: _(shim.isMac() ? 'Preferences' : 'Options'),
```
No conditional within _() as translation won't work

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+
+// Some commands aren't actually registered in CommandService
+// These hardcoded labels are used for those commands instead
+const commandLabels: CommandLabels = {
```
Do the labels get updated when the user changes the language ?

anjulalk replied:
I wrapped this logic in a function. Now it's working even after changing the language.

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+			keymapService.saveKeymap(); // Save changes to the disk
+
+			// TODO: Use events instead?
+			keymapService.triggerMenuRefresh();
```
Yes keymapService should make use of the events and dispatch events when the keymap is saved. The component should not dispatch any event

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+	};
+
+	return (
+		<>
```
Please wrap in a div

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+		);
+	};
+
+	const renderErrorMessage = (errorMessage: string) => {
```
No need to pass the errorMessage param since it’s part of the component state

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+
+			<div style={styles.container}>
+				<div style={styles.topActions}>
+					<input
```
Please add a Search label for accessibility purposes.

laurent22 replied:
Looks like the Search label hasn't been added yet?

laurent22 commented on ElectronClient/gui/KeymapConfig/ShortcutRecorder.tsx:
```diff
@@ -0,0 +1,65 @@
+const { _ } = require('lib/locale.js');
+const keymapService = KeymapService.instance();
+
+export interface ShortcutRecorderProps extends KeymapConfigScreenProps {
```
Please don’t use extend, just list all the props

laurent22 commented on ElectronClient/gui/KeymapConfig/ShortcutRecorder.tsx:
```diff
@@ -0,0 +1,65 @@
+			</button>
+
+			<button style={styles.inlineButton} onClick={handleReset}>
+				{_('Reset to Default')}
```
default lowercase

laurent22 commented on ReactNativeClient/lib/services/KeymapService.ts:
```diff
@@ -131,6 +137,17 @@ export default class KeymapService extends BaseService {
+
+		try {
+			const customKeymap = this.generateCustomKeymap();
+			await fs.writeFile(this.keymapPath, JSON.stringify(customKeymap, null, 2));
```
Use shim.fsDriver for file operations

laurent22 commented on ReactNativeClient/lib/services/KeymapService.ts:
```diff
@@ -168,11 +192,39 @@ export default class KeymapService extends BaseService {
+		return Object.keys(this.keymap);
+	}
+
+	setMenuRefreshCallback(callback: Function) {
```
Use eventEmitter

laurent22 commented on ReactNativeClient/lib/services/KeymapService.ts:
```diff
@@ -168,11 +192,39 @@ export default class KeymapService extends BaseService {
+		return Object.values(this.keymap);
+	}
+
+	getCommands() {
```
getCommandNames

laurent22 commented on ElectronClient/gui/KeymapConfig/ShortcutRecorder.tsx:
```diff
@@ -0,0 +1,65 @@
+	toggleEditing: () => void,
+}
+
+export const ShortcutRecorder = ({ setAccelerator, resetAccelerator, toggleEditing, theme }: ShortcutRecorderProps) => {
```
The child component logic should be decoupled from that of the parent. ShortcutRecorder should only take three event handlers: onSave, onCancel, onReset then on the parent component you handle the business logic to set keymap.

In fact I don’t see anything in this component to cancel input but that should be present

anjulalk replied:
I've updated the logic to use handlers instead. Thanks for pointing that out.

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+}
+
+export interface KeymapConfigScreenProps {
+	theme: number
```
Please name it themeId for compatibility with the coming ui update

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,168 @@
+			newEditing[commandName] = !newEditing[commandName];
+			return newEditing;
+		});
+	};
```
Regarding these three functions above, could you refactor using useReducer? I believe that’s really how it should be done in this case because a setter that calls a setter will be hard to maintain long term.

laurent22 replied:
Actually ignore this as once you’ll have refactored ShortcutRecorder you won’t need this layer of indirection, with setters calling setters, anymore.

laurent22 commented:
@anjulalk, any update on your pull request? If there's something tricky about some of my comments, please let me know.

anjulalk commented:
I'm still working on some issues. I'll be able to push the changes within the next 24 hours. Apologies for the delay.

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,144 @@
+
+		if ((invalidAccelerator && invalidKeymapItem.command === command) ||
+			(duplicateAccelerators && (duplicateKeymapItems[0].command === command || duplicateKeymapItems[1].command === command))) {
+			return '❌';
```
This will render differently and sometimes strangely depending on the operating system so please use an icon (Font Awesome is always loaded).

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,144 @@
+				<div style={styles.warning}>
+					<p style={styles.text}>
+						<span>
+							{error.details.duplicateAccelerators
+								? _('Keymap configuration contains duplicates. Change or disable one of the shortcuts to continue.')
+								: error.details.invalidAccelerator
+									? _('Keymap configuration contains an invalid shortcut. Change or disable it to continue.')
+									: error.message // Highly unlikely that any other error will occur at this point
+							}
```
These error messages should be part of the error object (in `error.message`), then here you can print error.message directly.

laurent22 commented on ReactNativeClient/lib/services/KeymapService.ts:
```diff
@@ -75,6 +78,36 @@ const defaultKeymap = {
+	duplicateKeymapItems?: [KeymapItem, KeymapItem]
+}
+
+export class KeymapError extends Error {
+	private details_: KeymapErrorDetails;
+
+	constructor(message: string, details: KeymapErrorDetails = {}) {
+		super(message);
+		this.details_ = {
+			missingAccelerator: false,
+			missingCommand: false,
+			invalidAccelerator: false,
+			invalidCommand: false,
+			duplicateAccelerators: false,
+			...details,
+		};
+	}
+
+	get details() {
+		return this.details_;
+	}
```
Could you extend JoplinError and simply set the code to some string? I think the details object with potentially multiple error causes seem unnecessarily complex. The keymap shortcut should be validated when the user inputs it anyway so we shouldn't have so much complexity going on. Check what they type, throw back an error with a message and that's almost all we need.

anjulalk replied:
I see. In that case I'll add some error handling to the ShortcutRecorder component. I'll update it.

anjulalk replied:
Is it okay to keep just an array of invalid Keymap items, so that we can show the little status icon next to the shortcut?

laurent22 replied:
What could cause the keymap to have multiple invalid items? If an item is invalid it shouldn't be possible to save it anyway, so there could only be one invalid item (the currently active, but unsaved, one), or am I missing something?

anjulalk replied:
No, you're right. We should prevent the user from causing the error by validating the shortcut before the user sets it. 
I'm working on it at the moment. I'll push an update in the next 24 hours.

laurent22 commented on ReactNativeClient/lib/services/KeymapService.ts:
```diff
@@ -75,6 +79,20 @@ const defaultKeymap = {
 	],
 };
 
+export class KeymapError extends Error {
+	// Translated error message for the GUI
+	private altMessage_: string;
+
+	constructor(message: string, altMessage: string) {
+		super(message);
+		this.altMessage_ = altMessage;
+	}
+
+	get altMessage() {
+		return this.altMessage_;
+	}
+}
+
```
Please remove the altMessage property - error should only have one message. In fact, as far as I can see there's no actual use for the KeymapError object anymore so please remove KeymapError and use a regular Error object.

To throw the error, please use this:

```js
new Error(_('Accelerator "%s" is already used for command "%s".', item.accelerator, originalItem.command));
```

anjulalk replied:
Thanks for mentioning the example. I agree that it's unneccessary. My initial idea was give a friendly error message in the GUI. 
I'll replace them with one message.

laurent22 reviewed (changes requested):
@anjulalk, thanks for the recent changes. I found just two remaining issues and I think it will be ready.

anjulalk commented:
I've added the label for the search input. Previously I thought you meant the `aria-label` property, so I added it. Sorry about that.

anjulalk commented:
I've pushed an update with Import/Export functionality. Please review when you've got the time. I've tested every feature and everything seems to be working as expected.

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,191 @@
+		if (filePath) {
+			try {
+				// KeymapService is already synchronized with the in-state keymap
+				await exportCustomKeymap(filePath);
```
Please call `await keymapService.saveCustomKeymap(filePath);` directly from here. You're importing KeymapService into this component anyway so the additional layer of indirection is unnecessary. Also it's best to keep React hooks small. So if you really wanted to use a hook here, it would be best to create a new one like "useImportExport" and put the import/export feature there, but it won't be necessary to do it now.

anjulalk replied:
I agree, it's unneccessary. I'll remove it and call the method directly.

laurent22 commented on ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx:
```diff
@@ -0,0 +1,191 @@
+				// KeymapService is already synchronized with the in-state keymap
+				await exportCustomKeymap(filePath);
+			} catch (err) {
+				bridge().showErrorMessageBox(`${_('An unexpected error occured while exporting the keymap!')}\n${err.message}`);
```
As an error here is unlikely, let's not create a new translatable string for it. So just `bridge().showErrorMessageBox(err.message);` for now would work.

anjulalk replied:
Sure, I'll remove it right away.

laurent22 reviewed (changes requested):
Thanks for the update @anjulalk, I've tried the editor and it's all looking good. I've only left two comments for the import/export feature.

Also please could you add the following tooltips?

- When a shortcut is invalid (for example for a duplicate shortcut), you display a small warning icon to the right, which is good, but please could you also add a tooltip on this icon with the complete error message (which would be from the error object that you throw)?

- Please could you add a tooltip over the shortcut recorder input field? It should be "Press the shortcut, then press ENTER. Or BACKSPACE to clear the shortcut".

anjulalk commented:
I made the requested changes base on your feedback. Please have a look.

After adding the label, it looks a bit odd because the label is "Search" and the search box also says "Search...". Can we change the search box placeholder to something like "Type in here to search...", or something like that? 

![image](https://user-images.githubusercontent.com/26695184/91811830-c03bc180-ec4d-11ea-863f-285b2650afc6.png)

IMHO, I think it'd look better without the label. But the current layout is fine too.

laurent22 commented:
Thanks for the update @anjulalk.

> After adding the label, it looks a bit odd because the label is "Search" and the search box also says "Search...". 

Hmm, now that you say it, it indeed doesn't look right. Could you remove it actually and make it like it was before? Sorry about that!

Also you have some conflicts on the ignore files. Once these small changes are done I think it's all good, and we can merge.

anjulalk commented:
I've updated the pull request. Please review when you have the time. Thanks!

laurent22 commented:
That's great, thanks @anjulalk! We'll make the feature part of the next release.
